src/bin/scrolls/daemon.rs

```diff
@@ -67,67 +48,141 @@ impl ConfigRoot {
+            gasket::runtime::TetherState::Alive(x) => {
+                matches!(x, gasket::runtime::StagePhase::Ended)
+            }
             _ => true,
         })
+    }
+
+    fn shutdown(&self) {
+        for tether in self.all_tethers() {
+            let state = tether.check_state();
+            warn!("dismissing stage: {} with state {:?}", tether.name(), state);
+            tether.dismiss_stage().expect("stage stops");
+
+            // Can't join the stage because there's a risk of deadlock, usually
+            // because a stage gets stuck sending into a port which depends on a
+            // different stage not yet dismissed. The solution is to either
+            // create a DAG of dependencies and dismiss in the
+            // correct order, or implement a 2-phase teardown where
+            // ports are disconnected and flushed before joining the
+            // stage.
+
+            //tether.join_stage();
+        }
+    }
 }
 
-fn shutdown(pipeline: bootstrap::Pipeline) {
-    for tether in pipeline.tethers {
-        let state = tether.check_state();
-        log::warn!("dismissing stage: {} with state {:?}", tether.name(), state);
-        tether.dismiss_stage().expect("stage stops");
+fn define_gasket_policy(config: Option<&gasket::retries::Policy>) -> gasket::runtime::Policy {
+    let default_policy = gasket::retries::Policy {
+        max_retries: 20,
+        backoff_unit: Duration::from_secs(1),
+        backoff_factor: 2,
+        max_backoff: Duration::from_secs(60),
+        dismissible: false,
+    };
+
+    gasket::runtime::Policy {
+        tick_timeout: None,
+        bootstrap_retry: config.cloned().unwrap_or(default_policy.clone()),
+        work_retry: config.cloned().unwrap_or(default_policy.clone()),
+        teardown_retry: config.cloned().unwrap_or(default_policy.clone()),
+    }
+}
 
-        // Can't join the stage because there's a risk of deadlock, usually
-        // because a stage gets stuck sending into a port which depends on a
-        // different stage not yet dismissed. The solution is to either create a
-        // DAG of dependencies and dismiss in the correct order, or implement a
-        // 2-phase teardown where ports are disconnected and flushed
-        // before joining the stage.
+fn chain_stages<'a>(
+    source: &'a mut dyn StageBootstrapper,
+    enrich: &'a mut dyn StageBootstrapper,
+    reducer: &'a mut dyn StageBootstrapper,
+    storage: &'a mut dyn StageBootstrapper,
+) {
+    let (to_next, from_prev) = gasket::messaging::tokio::broadcast_channel(100);
+    source.connect_output(to_next);
+    enrich.connect_input(from_prev);
+
+    let (to_next, from_prev) = gasket::messaging::tokio::broadcast_channel(100);
+    enrich.connect_output(to_next);
+    reducer.connect_input(from_prev);
+
+    let (to_next, from_prev) = gasket::messaging::tokio::broadcast_channel(100);
+    reducer.connect_output(to_next);
+    storage.connect_input(from_prev);
```

Comment:
we don't need broadcast for this, we can use mpsc

Reply:
Right. I changed it. Thanks <3